File: backend/app/api/audit_logs.py
from fastapi import APIRouter, Depends, Query, HTTPException, status

from app.core.auth import require_admin
from app.core.supabase import supabase


router = APIRouter(
    prefix="/api/audit-logs",
    tags=["Audit Logs"],
)


@router.get("")
def list_audit_logs(
    limit: int = Query(default=100, ge=1, le=200),
    current_user: dict = Depends(require_admin),
):
    try:
        response = (
            supabase.table("audit_logs")
            .select(
                "id, action, entity_type, entity_id, created_at, "
                "users(id, name, role)"
            )
            .order("created_at", desc=True)
            .limit(limit)
            .execute()
        )

        return response.data or []

    except Exception as exc:
        print(f"Audit log error: {exc}")

        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Unable to load audit log.",
        )

File: frontend/lib/api/audit-logs.ts
import { createClient } from "@/lib/supabase/server";

export interface AuditLog {
  id: string;
  action: string;
  entity_type: string;
  entity_id: string | null;
  created_at: string;
  users: {
    id: string;
    name: string;
    role: string;
  } | null;
}

export async function getAuditLogs(): Promise<
  AuditLog[]
> {
  const supabase = await createClient();

  const {
    data: { session },
  } = await supabase.auth.getSession();

  if (!session?.access_token) {
    throw new Error("Authentication required.");
  }

  const apiUrl =
    process.env.NEXT_PUBLIC_API_URL;

  if (!apiUrl) {
    throw new Error(
      "API URL is not configured.",
    );
  }

  const response = await fetch(
    `${apiUrl}/api/audit-logs?limit=100`,
    {
      headers: {
        Authorization: `Bearer ${session.access_token}`,
      },
      cache: "no-store",
    },
  );

  if (!response.ok) {
    throw new Error(
      "Unable to load audit logs.",
    );
  }

  return response.json();
}